// @cobalt/css-card/src/storybook.js
import React from "react";
import "./styles.scss";

const innerContentTemplate = (title, styles = { padding: "24px" }) => {
  return (
    <div style={styles}>
      <h2>{title}</h2>
      <p>
        {`Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Aenean
        commodo ligula eget dolor. Aenean massa. Cum sociis natoque penatibus
        et magnis dis parturient montes, nascetur ridiculus mus. Donec quam
        felis, ultricies nec, pellentesque eu, pretium quis, sem. Nulla
        consequat massa quis enim. Donec pede justo, fringilla vel, aliquet
        nec, vulputate eget, arcu. In enim justo, rhoncus ut, imperdiet a,
        venenatis vitae, justo. Nullam dictum felis eu pede mollis pretium.
        Integer tincidunt. Cras dapibus. Vivamus elementum semper nisi. Aenean
        vulputate eleifend tellus. Aenean leo ligula, porttitor eu, consequat
        vitae, eleifend ac, enim. Aliquam lorem ante, dapibus in, viverra
        quis, feugiat a, tellus. Phasellus viverra nulla ut metus varius
        laoreet. Quisque rutrum. Aenean imperdiet. Etiam ultricies nisi vel
        augue. Curabitur ullamcorper ultricies nisi. Nam eget dui. Etiam
        rhoncus. Maecenas tempus, tellus eget condimentum rhoncus, sem quam
        semper libero, sit amet adipiscing sem neque sed ipsum. Nam quam nunc,
        blandit vel, luctus pulvinar, hendrerit id, lorem. Maecenas nec odio
        et ante tincidunt tempus. Donec vitae sapien ut libero venenatis
        faucibus. Nullam quis ante. Etiam sit amet orci eget eros faucibus
        tincidunt. Duis leo. Sed fringilla mauris sit amet nibh. Donec sodales
        sagittis magna.`}
      </p>
    </div>
  );
};

const template = (
  modifiers = "",
  title = "Basic Card",
  styles = { padding: "24px" }
) => {
  return (
    <section className={`co-card ${modifiers}`}>
      {innerContentTemplate(title, styles)}
    </section>
  );
};

export const basic = () => {
  return template();
};

export const outlined = () => {
  return template("co-card--outlined", "Outlined Card");
};

export const clickable = () => {
  return template("co-card--clickable", "Clickable Card");
};

export const withTopAndBottomContent = () => {
  return (
    <section className="co-card">
      {innerContentTemplate("Top Content", {
        padding: "24px",
        backgroundColor: "var(--primary-200)"
      })}
      {innerContentTemplate("Top Content", {
        padding: "24px",
        backgroundColor: "var(--gray-300)"
      })}
    </section>
  );
};

export default { title: "Card" };
