// @cobalt/css-button/src/storybook.js
import React from "react";
import "./styles.scss";

export const filled = () => {
  return (
    <div
      style={{ display: "flex", alignItems: "center", flexDirection: "row" }}
    >
      <button
        type="button"
        className="co-button co-button--primary"
        style={{ margin: "4px" }}
      >
        <span>Primary</span>
      </button>
      <button
        type="button"
        className="co-button co-button--secondary"
        style={{ margin: "4px" }}
      >
        <span>Secondary</span>
      </button>
      <button
        type="button"
        className="co-button co-button--danger"
        style={{ margin: "4px" }}
      >
        <span>Danger</span>
      </button>
      <button
        type="button"
        className="co-button co-button--warning"
        style={{ margin: "4px" }}
      >
        <span>Warning</span>
      </button>
      <button
        type="button"
        className="co-button co-button--success"
        style={{ margin: "4px" }}
      >
        <span>Success</span>
      </button>
      <button
        type="button"
        className="co-button co-button--feature"
        style={{ margin: "4px" }}
      >
        <span>Feature</span>
      </button>
      <button
        type="button"
        className="co-button co-button--primary"
        disabled
        style={{ margin: "4px" }}
      >
        <span>Disabled</span>
      </button>
    </div>
  );
};

export const outline = () => {
  return (
    <div
      style={{ display: "flex", alignItems: "center", flexDirection: "row" }}
    >
      <button
        type="button"
        className="co-button co-button--outline co-button--primary"
        style={{ margin: "4px" }}
      >
        <span>Primary</span>
      </button>
      <button
        type="button"
        className="co-button co-button--outline co-button--secondary"
        style={{ margin: "4px" }}
      >
        <span>Secondary</span>
      </button>
      <button
        type="button"
        className="co-button co-button--outline co-button--danger"
        style={{ margin: "4px" }}
      >
        <span>Danger</span>
      </button>
      <button
        type="button"
        className="co-button co-button--outline co-button--warning"
        style={{ margin: "4px" }}
      >
        <span>Warning</span>
      </button>
      <button
        type="button"
        className="co-button co-button--outline co-button--success"
        style={{ margin: "4px" }}
      >
        <span>Success</span>
      </button>
      <button
        type="button"
        className="co-button co-button--outline co-button--feature"
        style={{ margin: "4px" }}
      >
        <span>Feature</span>
      </button>
      <button
        type="button"
        className="co-button co-button--outline co-button--primary"
        disabled
        style={{ margin: "4px" }}
      >
        <span>Disabled</span>
      </button>
    </div>
  );
};

export const transparent = () => {
  return (
    <div
      style={{ display: "flex", alignItems: "center", flexDirection: "row" }}
    >
      <button
        type="button"
        className="co-button co-button--transparent co-button--primary"
        style={{ margin: "4px" }}
      >
        <span>Primary</span>
      </button>
      <button
        type="button"
        className="co-button co-button--transparent co-button--secondary"
        style={{ margin: "4px" }}
      >
        <span>Secondary</span>
      </button>
      <button
        type="button"
        className="co-button co-button--transparent co-button--danger"
        style={{ margin: "4px" }}
      >
        <span>Danger</span>
      </button>
      <button
        type="button"
        className="co-button co-button--transparent co-button--warning"
        style={{ margin: "4px" }}
      >
        <span>Warning</span>
      </button>
      <button
        type="button"
        className="co-button co-button--transparent co-button--success"
        style={{ margin: "4px" }}
      >
        <span>Success</span>
      </button>
      <button
        type="button"
        className="co-button co-button--transparent co-button--feature"
        style={{ margin: "4px" }}
      >
        <span>Feature</span>
      </button>
      <button
        type="button"
        className="co-button co-button--transparent co-button--primary"
        disabled
        style={{ margin: "4px" }}
      >
        <span>Disabled</span>
      </button>
    </div>
  );
};

export const monochrome = () => {
  return (
    <div style={{ display: "flex" }}>
      <div
        style={{
          display: "flex",
          alignItems: "center",
          padding: 8,
          border: "2px solid var(--black)",
          borderRight: 0,
          backgroundColor: "var(--white)"
        }}
      >
        <button
          type="button"
          className="co-button co-button--mono-dark"
          style={{ margin: "4px" }}
        >
          <span>Mono Dark</span>
        </button>
        <button
          type="button"
          className="co-button co-button--outline co-button--mono-dark"
          style={{ margin: "4px" }}
        >
          <span>Mono Dark</span>
        </button>
        <button
          type="button"
          className="co-button co-button--transparent co-button--mono-dark"
          style={{ margin: "4px" }}
        >
          <span>Mono Dark</span>
        </button>
      </div>
      <div
        style={{
          display: "flex",
          alignItems: "center",
          padding: 8,
          border: "2px solid var(--black)",
          borderLeft: 0,
          backgroundColor: "var(--black)"
        }}
      >
        <button
          type="button"
          className="co-button co-button--mono-light"
          style={{ margin: "4px" }}
        >
          <span>Mono Light</span>
        </button>
        <button
          type="button"
          className="co-button co-button--outline co-button--mono-light"
          style={{ margin: "4px" }}
        >
          <span>Mono Light</span>
        </button>
        <button
          type="button"
          className="co-button co-button--transparent co-button--mono-light"
          style={{ margin: "4px" }}
        >
          <span>Mono Light</span>
        </button>
      </div>
    </div>
  );
};

export const sizes = () => {
  return (
    <div
      style={{ display: "flex", alignItems: "center", flexDirection: "row" }}
    >
      <button
        type="button"
        className="co-button co-button--primary co-button--small"
        style={{ margin: "4px" }}
      >
        <span>Small</span>
      </button>
      <button
        type="button"
        className="co-button co-button--primary co-button--medium"
        style={{ margin: "4px" }}
      >
        <span>Medium</span>
      </button>
    </div>
  );
};

export const Shape = () => {
  return (
    <>
      <div
        style={{ display: "flex", alignItems: "center", flexDirection: "row" }}
      >
        <button
          type="button"
          className="co-button co-button--primary co-button--compact"
          style={{ margin: "4px" }}
        >
          <span
            style={{ width: "18px", height: "18px", background: "#fff" }}
          ></span>
        </button>
        <button
          type="button"
          className="co-button co-button--primary co-button--full-width"
          style={{ margin: "4px" }}
        >
          <span>Full width</span>
        </button>
      </div>
    </>
  );
};

export default { title: "Button" };
