// @cobalt/css-avatar/src/storybook.js
import React from "react";
import "./styles.scss";

export const basic = () => {
  return (
    <div className="co-avatar">
      <div className="co-avatar__content">CO</div>
    </div>
  );
};

export const sizes = () => {
  return (
    <div>
      <div className="co-avatar co-avatar--tiny" style={{ marginRight: "8px" }}>
        <div className="co-avatar__content">XS</div>
      </div>
      <div
        className="co-avatar co-avatar--small"
        style={{ marginRight: "8px" }}
      >
        <div className="co-avatar__content">S</div>
      </div>
      <div
        className="co-avatar co-avatar--medium"
        style={{ marginRight: "8px" }}
      >
        <div className="co-avatar__content">M</div>
      </div>
      <div className="co-avatar co-avatar--large">
        <div className="co-avatar__content">L</div>
      </div>
    </div>
  );
};

export const colors = () => {
  return (
    <div>
      <div className="co-avatar" style={{ marginRight: "8px" }}>
        <div className="co-avatar__content">CO</div>
      </div>
      <div
        className="co-avatar co-avatar--theme"
        style={{ marginRight: "8px" }}
      >
        <div className="co-avatar__content">TH</div>
      </div>
      <div
        className="co-avatar co-avatar--light-theme"
        style={{ marginRight: "8px" }}
      >
        <div className="co-avatar__content">LT</div>
      </div>
      <div className="co-avatar co-avatar--blue" style={{ marginRight: "8px" }}>
        <div className="co-avatar__content">B</div>
      </div>
      <div
        className="co-avatar co-avatar--green"
        style={{ marginRight: "8px" }}
      >
        <div className="co-avatar__content">G</div>
      </div>
      <div
        className="co-avatar co-avatar--purple"
        style={{ marginRight: "8px" }}
      >
        <div className="co-avatar__content">P</div>
      </div>
      <div className="co-avatar co-avatar--teal" style={{ marginRight: "8px" }}>
        <div className="co-avatar__content">T</div>
      </div>
      <div
        className="co-avatar co-avatar--yellow"
        style={{ marginRight: "8px" }}
      >
        <div className="co-avatar__content">Y</div>
      </div>
      <div className="co-avatar co-avatar--red" style={{ marginRight: "8px" }}>
        <div className="co-avatar__content">R</div>
      </div>
      <div className="co-avatar co-avatar--orange">
        <div className="co-avatar__content">O</div>
      </div>
    </div>
  );
};

export const status = () => {
  return (
    <div>
      <div className="co-avatar co-avatar--status-gray" style={{ marginRight: "8px" }}>
        <div className="co-avatar__content">K</div>
      </div>
      <div className="co-avatar co-avatar--status-green" style={{ marginRight: "8px" }}>
        <div className="co-avatar__content">G</div>
      </div>
      <div className="co-avatar co-avatar--status-yellow" style={{ marginRight: "8px" }}>
        <div className="co-avatar__content">Y</div>
      </div>
      <div className="co-avatar co-avatar--status-red">
        <div className="co-avatar__content">R</div>
      </div>
    </div>
  );
};

export const skeleton = () => {
  return (
    <div style={{ display: "flex", flexDirection: "column" }}>
      <p>Static</p>
      <div className="co-avatar co-avatar--skeleton">
        <div className="co-avatar__content"></div>
      </div>
      <p>Animated</p>
      <div className="co-avatar co-avatar--animated-skeleton">
        <div className="co-avatar__content"></div>
      </div>
    </div>
  );
};

export const border = () => {
  return (
    <div
      style={{
        width: "124px",
        height: "52px",
        padding: "4px",
        backgroundColor: "gray",
        "--background-color": "gray"
      }}
    >
      <div className="co-avatar">
        <div className="co-avatar__content">C</div>
      </div>
      <div
        style={{ position: "relative", marginLeft: "-16px" }}
        className="co-avatar"
      >
        <div className="co-avatar__content">C</div>
      </div>
      <div
        style={{ position: "relative", marginLeft: "-16px" }}
        className="co-avatar"
      >
        <div className="co-avatar__content">C</div>
      </div>
    </div>
  );
};

export default { title: "Avatar" };
