// @cobalt/css-panel/src/storybook.js
import React from "react";
import "./styles.scss";

export const usage = () => (
  <div className="co-panel co-panel--medium co-panel--right co-panel--visible">
    <div
      style={{
        display: "flex",
        justifyContent: "center",
        marginTop: "20px"
      }}
    >
      <h2>Panel</h2>
    </div>
  </div>
);

export const small = () => (
  <div className="co-panel co-panel--small co-panel--right co-panel--visible">
    <div
      style={{
        display: "flex",
        justifyContent: "center",
        marginTop: "20px"
      }}
    >
      <h2>Small Panel</h2>
    </div>
  </div>
);

export const large = () => (
  <div className="co-panel co-panel--large co-panel--right co-panel--visible">
    <div
      style={{
        display: "flex",
        justifyContent: "center",
        marginTop: "20px"
      }}
    >
      <h2>Large Panel</h2>
    </div>
  </div>
);

export const left = () => (
  <div className="co-panel co-panel--medium co-panel--left co-panel--visible">
    <div
      style={{
        display: "flex",
        justifyContent: "center",
        marginTop: "20px"
      }}
    >
      <h2>Left Panel</h2>
    </div>
  </div>
);

export const overlay = () => (
  <>
    <div className="co-panel__overlay"></div>
    <div className="co-panel co-panel--medium co-panel--right co-panel--visible">
      <div
        style={{
          display: "flex",
          justifyContent: "center",
          marginTop: "20px"
        }}
      >
        <h2>Overlay</h2>
      </div>
    </div>
  </>
);

export default { title: "Panel" };
