// @cobalt/css-flex/src/storybook.js
import React from "react";
import "./styles.scss";

const Circle = ({ children, small }) => (
  <div
    style={{
      display: "flex",
      alignItems: "center",
      justifyContent: "center",
      width: small ? 50 : 100,
      height: small ? 50 : 100,
      backgroundColor: "#EEE",
      boxSizing: "border-box",
      border: "1px solid #AAA",
      borderRadius: "50%"
    }}
  >
    {children}
  </div>
);
const Rect = ({ children, small }) => (
  <div
    style={{
      display: "flex",
      alignItems: "center",
      justifyContent: "center",
      width: small ? 50 : 100,
      height: small ? 50 : 100,
      backgroundColor: "#EEE",
      boxSizing: "border-box",
      border: "1px solid #AAA"
    }}
  >
    {children}
  </div>
);

export const primary = () => {
  return (
    <div className="co-flex co-flex--row">
      <Rect />
      <Rect />
    </div>
  );
};

export const column = () => {
  return (
    <div className="co-flex co-flex--column">
      <Rect />
      <Rect />
    </div>
  );
};

export const padding = () => {
  return (
    <>
      <div className="co-flex co-flex--row co-flex--padding-2">
        <Rect />
      </div>
      <div className="co-flex co-flex--row co-flex--padding-x-4 co-flex--padding-y-4">
        <Rect />
      </div>
      <div className="co-flex co-flex--row co-flex--padding-left-6 co-flex--padding-right-6 co-flex--padding-top-6 co-flex--padding-bottom-6">
        <Rect />
      </div>
      <div className="co-flex co-flex-row co-flex--padding-2 co-flex--padding-y-4 co-flex--padding-top-8">
        <Rect />
      </div>
    </>
  );
};

export const gap = () => {
  return (
    <>
      {[1, 2, 3, 4, 5, 6, 7, 8].map(v => (
        <div
          className={`co-flex co-flex--row co-flex--gap-${v}`}
          style={{ padding: "8px 0" }}
        >
          {new Array(4).fill(0).map((v, i) => (
            <div className="co-flex__item">
              <Rect>{i}</Rect>
            </div>
          ))}
        </div>
      ))}
    </>
  );
};

export const inwards = () => {
  return (
    <div className="co-flex co-flex--row co-flex--gap-8">
      <div className="co-flex__item">
        <div
          className="co-flex co-flex--row co-flex--gap-8 co-flex--inwards"
          style={{ padding: "8px 0" }}
        >
          {new Array(4).fill(0).map((v, i) => (
            <div className="co-flex__item">
              <Circle>{i}</Circle>
            </div>
          ))}
        </div>
      </div>
      <div className="co-flex__item">
        <div
          className="co-flex co-flex--column co-flex--gap-8 co-flex--inwards"
          style={{ padding: "8px 0" }}
        >
          {new Array(4).fill(0).map((v, i) => (
            <div className="co-flex__item">
              <Circle>{i}</Circle>
            </div>
          ))}
        </div>
      </div>
    </div>
  );
};

export const wrap = () => {
  return (
    <div className="co-flex co-flex--row">
      <div style={{ width: 500, height: 500 }}>
        <div className="co-flex co-flex--row co-flex--gap-2 co-flex--wrap">
          {new Array(12).fill(0).map((v, i) => (
            <div className="co-flex__item">
              <Rect>{i}</Rect>
            </div>
          ))}
        </div>
      </div>
      <div style={{ width: 500, height: 500 }}>
        <div className="co-flex co-flex--column co-flex--gap-2 co-flex--wrap">
          {new Array(12).fill(0).map((v, i) => (
            <div className="co-flex__item">
              <Rect>{i}</Rect>
            </div>
          ))}
        </div>
      </div>
    </div>
  );
};

export const inwardsWrap = () => {
  return (
    <div className="co-flex co-flex--row">
      <div style={{ width: 500, height: 500 }}>
        <div className="co-flex co-flex--row co-flex--gap-8 co-flex--wrap co-flex--inwards">
          {new Array(12).fill(0).map((v, i) => (
            <div className="co-flex__item">
              <Rect>{i}</Rect>
            </div>
          ))}
        </div>
      </div>
      <div style={{ width: 500, height: 500 }}>
        <div className="co-flex co-flex--column co-flex--gap-8 co-flex--wrap co-flex--inwards">
          {new Array(12).fill(0).map((v, i) => (
            <div className="co-flex__item">
              <Rect>{i}</Rect>
            </div>
          ))}
        </div>
      </div>
    </div>
  );
};

export const nestedGaps = () => (
  <div className="co-flex co-flex--row co-flex--gap-8">
    <div className="co-flex__item">
      <Rect />
    </div>
    <div className="co-flex__item">
      <div className="co-flex co-flex--row co-flex--gap-4">
        <div className="co-flex__item">
          <Rect />
        </div>
        <div className="co-flex__item">
          <div className="co-flex co-flex--row co-flex--gap-2">
            <div className="co-flex__item">
              <Rect />
            </div>
            <div className="co-flex__item">
              <Rect />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
);

export const growing = () => (
  <div className="co-flex co-flex--row">
    <div className="co-flex co-flex--row">
      <div className="co-flex co-flex--row co-flex--gap-8">
        <div className="co-flex__item">
          <Rect />
        </div>
        <div className="co-flex__item">
          <Rect />
        </div>
      </div>
    </div>
    <div className="co-flex co-flex--row co-flex--grow">
      <div className="co-flex co-flex--row co-flex--gap-2 co-flex--align-x-end co-flex--align-y-center">
        <div className="co-flex__item">
          <Rect small />
        </div>
        <div className="co-flex__item">
          <Rect />
        </div>
      </div>
    </div>
  </div>
);

export default { title: "Flex" };
