// @cobalt/css-spinner/src/storybook.js
import React from "react";
import "./styles.scss";

export const usage = () => {
  return <i className="co-spinner" />;
};

export const sizes = () => {
  return (
    <>
      <div
        style={{
          display: "flex",
          flexDirection: "row",
          alignItems: "center",
          marginBottom: "8px"
        }}
      >
        <i className="co-spinner co-spinner--micro" />
        <p style={{ margin: "0 8px" }}>Micro</p>
      </div>
      <div
        style={{
          display: "flex",
          flexDirection: "row",
          alignItems: "center",
          marginBottom: "8px"
        }}
      >
        <i className="co-spinner co-spinner--tiny" />
        <p style={{ margin: "0 8px" }}>Tiny</p>
      </div>
      <div
        style={{
          display: "flex",
          flexDirection: "row",
          alignItems: "center",
          marginBottom: "8px"
        }}
      >
        <i className="co-spinner co-spinner--small" />
        <p style={{ margin: "0 8px" }}>Small</p>
      </div>
      <div
        style={{
          display: "flex",
          flexDirection: "row",
          alignItems: "center",
          marginBottom: "8px"
        }}
      >
        <i className="co-spinner co-spinner--medium" />
        <p style={{ margin: "0 8px" }}>Medium</p>
      </div>
      <div
        style={{
          display: "flex",
          flexDirection: "row",
          alignItems: "center",
          marginBottom: "8px"
        }}
      >
        <i className="co-spinner co-spinner--large" />
        <p style={{ margin: "0 8px" }}>Large</p>
      </div>
      <div
        style={{
          display: "flex",
          flexDirection: "row",
          alignItems: "center",
          marginBottom: "8px"
        }}
      >
        <i className="co-spinner co-spinner--xlarge" />
        <p style={{ margin: "0 8px" }}>XLarge</p>
      </div>
    </>
  );
};

export default { title: "Spinner" };
