// @cobalt/react-calendar/src/Calendar.tsx
import React, { FC, HTMLAttributes, Ref } from "react";
import cc from "classcat";
import {
  CalendarType,
  CallbackDate,
  DaysType,
  Month,
  Weekdays as WeekdaysType,
  Weekday
} from "./types";
import Weekdays from "./Weekdays";
import Days from "./Days";
import { validateDates } from "./utils";
import styles from "@cobalt/css-calendar";

export interface Props extends HTMLAttributes<CalendarType> {
  month: Month;
  weekdays: WeekdaysType;
  year: number | string;
  disabledDays?: DaysType;
  highlightedDays?: DaysType;
  onDayClick?: CallbackDate;
  onDayMouseEnter?: CallbackDate;
  selectedDays?: DaysType;
  weekStart?: Weekday;
  forwardedRef?: Ref<CalendarType>;
}

const Calendar: FC<Props> = ({
  className = "",
  forwardedRef,
  month,
  weekdays,
  year,
  disabledDays = [],
  highlightedDays = [],
  onDayClick,
  onDayMouseEnter,
  selectedDays = [],
  weekStart = "MON",
  ...props
}) => {
  const monthNr = Number(month);
  const yearNr = Number(year);

  validateDates(yearNr, monthNr);

  const classStyles = cc([styles["co-calendar"], className]);

  return (
    <div className={classStyles} ref={forwardedRef} {...props}>
      <Weekdays days={weekdays} weekStart={weekStart} />
      <Days
        month={monthNr}
        year={yearNr}
        disabledDays={disabledDays}
        highlightedDays={highlightedDays}
        onDayClick={onDayClick}
        onDayMouseEnter={onDayMouseEnter}
        selectedDays={selectedDays}
        weekStart={weekStart}
      />
    </div>
  );
};

export default Calendar;
