// @cobalt/react-calendar/src/Day.tsx
import React, { FC } from "react";
import cc from "classcat";
import DayInner from "./DayInner";
import { CallbackDate, Date, SelectionType } from "./types";
import styles from "@cobalt/css-calendar";

export interface Props {
  date: Date;
  disabled?: boolean;
  highlighted?: boolean;
  onClick?: CallbackDate;
  onMouseEnter?: CallbackDate;
  selection?: SelectionType;
}

const Day: FC<Props> = ({
  date,
  disabled,
  highlighted,
  onClick,
  onMouseEnter,
  selection
}) => {
  const classStyles = cc([
    styles["co-calendar__day"],
    {
      [styles["co-calendar__day--highlighted"]]: highlighted,
      [styles["co-calendar__day--selected"]]: selection,
      [styles["co-calendar__day--range"]]: selection && selection !== "single",
      [styles["co-calendar__day--range-first"]]: selection === "first",
      [styles["co-calendar__day--range-last"]]: selection === "last"
    }
  ]);

  return (
    <div
      className={classStyles}
      aria-disabled={disabled || undefined}
      data-co-disabled={disabled || undefined}
    >
      <DayInner date={date} onClick={onClick} onMouseEnter={onMouseEnter} />
    </div>
  );
};

export default Day;
