// @cobalt/react-calendar/src/Days.tsx
import React, { FC } from "react";
import styles from "@cobalt/css-calendar";
import Day from "./Day";
import { CallbackDate, DaysType, Weekday } from "./types";
import * as utils from "./utils";

export interface Props {
  month: number;
  year: number;
  weekStart: Weekday;
  disabledDays: DaysType;
  nrEmptyDays?: number;
  highlightedDays: DaysType;
  onDayClick?: CallbackDate;
  onDayMouseEnter?: CallbackDate;
  selectedDays: DaysType;
}

const Days: FC<Props> = ({
  month,
  year,
  disabledDays,
  weekStart,
  highlightedDays,
  onDayClick,
  onDayMouseEnter,
  selectedDays
}) => {
  const emptyDays = utils
    .getEmptyDays(weekStart, year, month)
    .map((_, index) => {
      return <div key={index} />;
    });

  const monthsDays = utils.getDaysInMonth(year, month).map((_, index) => {
    const day = index + 1;
    const currentDate = { year, month, day };

    const isDisabled = !!utils.containsDay(disabledDays, day);
    const isHighlighted = !!utils.containsDay(highlightedDays, day);
    const selection = utils.containsDay(selectedDays, day);

    return (
      <Day
        key={index}
        date={currentDate}
        disabled={isDisabled}
        highlighted={isHighlighted}
        onClick={onDayClick}
        onMouseEnter={onDayMouseEnter}
        selection={selection}
      />
    );
  });

  return (
    <div className={styles["co-calendar__days"]}>
      {emptyDays}
      {monthsDays}
    </div>
  );
};

export default Days;
