// @cobalt/react-calendar/src/types.ts
export type CalendarType = HTMLDivElement;

export type Weekday = "MON" | "TUE" | "WED" | "THU" | "FRI" | "SAT" | "SUN";

export type Weekdays = {
  [k in Weekday]: string;
};

export interface DaysRange {
  start?: Day;
  end?: Day;
}

export type DaysType = Array<DaysRange | Day>;

export type Date = {
  year: number;
  month: number;
  day: number;
};

export type CallbackDate = (date: Date) => void;

export type RangeType = "first" | "between" | "last" | undefined;

export type SelectionType = RangeType | "single";

export type Month =
  | 1
  | "1"
  | 2
  | "2"
  | 3
  | "3"
  | 4
  | "4"
  | 5
  | "5"
  | 6
  | "6"
  | 7
  | "7"
  | 8
  | "8"
  | 9
  | "9"
  | 10
  | "10"
  | 11
  | "11"
  | 12
  | "12";

export type Day =
  | 1
  | "1"
  | 2
  | "2"
  | 3
  | "3"
  | 4
  | "4"
  | 5
  | "5"
  | 6
  | "6"
  | 7
  | "7"
  | 8
  | "8"
  | 9
  | "9"
  | 10
  | "10"
  | 11
  | "11"
  | 12
  | "12"
  | 13
  | "13"
  | 14
  | "14"
  | 15
  | "15"
  | 16
  | "16"
  | 17
  | "17"
  | 18
  | "18"
  | 19
  | "19"
  | 20
  | "20"
  | 21
  | "21"
  | 22
  | "22"
  | 23
  | "23"
  | 24
  | "24"
  | 25
  | "25"
  | 26
  | "26"
  | 27
  | "27"
  | 28
  | "28"
  | 29
  | "29"
  | 30
  | "30"
  | 31
  | "31";
