// @cobalt/react-tab/src/Tab.tsx
import React, { FC, AnchorHTMLAttributes, Ref } from "react";
import cc from "classcat";
import styles from "@cobalt/css-tab";

type Size = "small" | "medium";
type Styles = typeof styles;

export type TabType = HTMLAnchorElement;

export interface Props extends AnchorHTMLAttributes<TabType> {
  active?: boolean;
  disabled?: boolean;
  size?: Size;
  forwardedRef?: Ref<TabType>;
}

const Tab: FC<Props> = ({
  className = "",
  forwardedRef,
  active,
  children,
  disabled,
  size = "medium",
  ...props
}) => {
  const classStyles = [
    styles["co-tab"],
    styles[`co-tab--${size}` as keyof Styles],
    {
      [styles["co-tab--active"]]: active
    },
    className
  ];
  return (
    <>
      {disabled ? (
        <span
          className={cc(classStyles)}
          ref={forwardedRef}
          data-co-disabled
          aria-disabled
          {...props}
        >
          {children}
        </span>
      ) : (
        <a className={cc(classStyles)} ref={forwardedRef} {...props}>
          {children}
        </a>
      )}
    </>
  );
};

export default Tab;
