// @cobalt/react-table/src/Cell.tsx
import React, { FC, HTMLAttributes, Ref } from "react";
import cc from "classcat";
import styles from "@cobalt/css-table";
import { Width } from "./types";

export type CellType = HTMLTableCellElement;
type Styles = typeof styles;

export interface Props extends HTMLAttributes<CellType> {
  as?: "td" | "div";
  /**
   * Defines the percentage of the width the cell should take.
   * Use values from 5 to 100, in multiples of 5.
   * 33 and 66 are also possible values.
   * To have the cell occupy the minimum space needed, use "min".
   */
  width?: Width;
  forwardedRef?: Ref<CellType>;
}

export const getCellWidth = (width: Width | undefined) =>
  width && styles[`co-table__cell--${width}` as keyof Styles];

const Cell: FC<Props> = ({
  className = "",
  forwardedRef,
  as = "td",
  width,
  ...props
}) => {
  const classes = cc([
    styles["co-table__cell"],
    getCellWidth(width),
    className
  ]);

  const Tag = as;

  return <Tag className={classes} ref={forwardedRef} {...props} />;
};

export default Cell;
