// @cobalt/css-table/src/storybook.js
import React from "react";
import "./styles.scss";

export const primary = () => {
  return (
    <table className="co-table">
      <thead className="co-table__head">
        <tr className="co-table__row">
          <th className="co-table__cell">Head cell</th>
          <th className="co-table__cell">Head cell</th>
        </tr>
      </thead>
      <tbody className="co-table__body">
        <tr className="co-table__row">
          <td className="co-table__cell">Table cell</td>
          <td className="co-table__cell">Table cell</td>
        </tr>
        <tr className="co-table__row">
          <td className="co-table__cell">Table cell</td>
          <td className="co-table__cell">Table cell</td>
        </tr>
      </tbody>
    </table>
  );
};

export const borderless = () => {
  return (
    <table className="co-table co-table--borderless">
      <thead className="co-table__head">
        <tr className="co-table__row">
          <th className="co-table__cell">Head cell</th>
          <th className="co-table__cell">Head cell</th>
        </tr>
      </thead>
      <tbody className="co-table__body">
        <tr className="co-table__row">
          <td className="co-table__cell">Table cell</td>
          <td className="co-table__cell">Table cell</td>
        </tr>
        <tr className="co-table__row">
          <td className="co-table__cell">Table cell</td>
          <td className="co-table__cell">Table cell</td>
        </tr>
      </tbody>
    </table>
  );
};

export const bordered = () => {
  return (
    <table className="co-table co-table--bordered">
      <thead className="co-table__head">
        <tr className="co-table__row">
          <th className="co-table__cell">Head cell</th>
          <th className="co-table__cell">Head cell</th>
        </tr>
      </thead>
      <tbody className="co-table__body">
        <tr className="co-table__row">
          <td className="co-table__cell">Table cell</td>
          <td className="co-table__cell">Table cell</td>
        </tr>
        <tr className="co-table__row">
          <td className="co-table__cell">Table cell</td>
          <td className="co-table__cell">Table cell</td>
        </tr>
      </tbody>
    </table>
  );
};

export const variation = () => {
  return (
    <table className="co-table">
      <thead className="co-table__head">
        <tr className="co-table__row">
          <th className="co-table__cell">Head cell</th>
          <th className="co-table__cell">Head cell</th>
        </tr>
      </thead>
      <tbody className="co-table__body">
        <tr className="co-table__row co-table__row--theme">
          <td className="co-table__cell">Table cell</td>
          <td className="co-table__cell">Table cell</td>
        </tr>
        <tr className="co-table__row co-table__row--success">
          <td className="co-table__cell">Table cell</td>
          <td className="co-table__cell">Table cell</td>
        </tr>
        <tr className="co-table__row co-table__row--warning">
          <td className="co-table__cell">Table cell</td>
          <td className="co-table__cell">Table cell</td>
        </tr>
        <tr className="co-table__row co-table__row--error">
          <td className="co-table__cell">Table cell</td>
          <td className="co-table__cell">Table cell</td>
        </tr>
        <tr className="co-table__row co-table__row--feature">
          <td className="co-table__cell">Table cell</td>
          <td className="co-table__cell">Table cell</td>
        </tr>
      </tbody>
    </table>
  );
};

export const widths = () => {
  return (
    <table className="co-table co-table--bordered">
      <thead className="co-table__head">
        <tr className="co-table__row">
          <th className="co-table__cell co-table__cell--30">Head cell</th>
          <th className="co-table__cell co-table__cell--70">Head cell</th>
        </tr>
      </thead>
      <tbody className="co-table__body">
        <tr className="co-table__row">
          <td className="co-table__cell co-table__cell--30">Table cell</td>
          <td className="co-table__cell co-table__cell--70">Table cell</td>
        </tr>
        <tr className="co-table__row">
          <td className="co-table__cell co-table__cell--30">Table cell</td>
          <td className="co-table__cell co-table__cell--70">Table cell</td>
        </tr>
      </tbody>
    </table>
  );
};

export const withDragHandle = () => (
  <table className="co-table">
    <thead className="co-table__head">
      <tr className="co-table__row">
        <th className="co-table__row-drag-handle co-table__row-drag-handle--placeholder" />
        <th className="co-table__cell co-table__cell--30">Head cell</th>
        <th className="co-table__cell co-table__cell--70">Head cell</th>
      </tr>
    </thead>
    <tbody className="co-table__body">
      <tr className="co-table__row">
        <td className="co-table__row-drag-handle">&#x2022;</td>
        <td className="co-table__cell co-table__cell--30">Table cell</td>
        <td className="co-table__cell co-table__cell--70">Table cell</td>
      </tr>
      <tr className="co-table__row">
        <td className="co-table__row-drag-handle">&#x2022;</td>
        <td className="co-table__cell co-table__cell--30">Table cell</td>
        <td className="co-table__cell co-table__cell--70">Table cell</td>
      </tr>
    </tbody>
  </table>
);

export const clickable = () => {
  return (
    <table className="co-table co-table--bordered">
      <thead className="co-table__head">
        <tr className="co-table__row">
          <th className="co-table__cell">Head cell</th>
          <th className="co-table__cell">Head cell</th>
        </tr>
      </thead>
      <tbody className="co-table__body">
        <tr className="co-table__row co-table__row--clickable">
          <td className="co-table__cell">Table cell</td>
          <td className="co-table__cell">Table cell</td>
        </tr>
        <tr className="co-table__row co-table__row--clickable">
          <td className="co-table__cell">Table cell</td>
          <td className="co-table__cell">Table cell</td>
        </tr>
      </tbody>
    </table>
  );
};

export const expandableRow = () => {
  return (
    <table className="co-table">
      <tbody className="co-table__body">
        <tr className="co-table__row co-table__row--clickable co-table__row--expanded">
          <td className="co-table__cell">Table cell</td>
          <td className="co-table__cell">Table cell</td>
        </tr>
        <tr className="co-table__row-content co-table__row-content--expanded">
          <td className="co-table__cell">
            This is the content of the expandable row
          </td>
        </tr>
      </tbody>
      <tbody className="co-table__body">
        <tr className="co-table__row co-table__row--clickable co-table__row--expandable">
          <td className="co-table__cell">Table cell</td>
          <td className="co-table__cell">Table cell</td>
        </tr>
        <tr className="co-table__row-content">
          <td className="co-table__cell">
            This is the content of the expandable row
          </td>
        </tr>
      </tbody>
    </table>
  );
};

export const usingDIV = () => {
  return (
    <div className="co-table co-table--bordered">
      <div className="co-table__head">
        <div className="co-table__row">
          <div className="co-table__cell">Head cell</div>
          <div className="co-table__cell">Head cell</div>
        </div>
      </div>
      <tbody className="co-table__body">
        <div className="co-table__row">
          <div className="co-table__cell">Table cell</div>
          <div className="co-table__cell">Table cell</div>
        </div>
        <div className="co-table__row">
          <div className="co-table__cell">Table cell</div>
          <div className="co-table__cell">Table cell</div>
        </div>
      </tbody>
    </div>
  );
};

export default { title: "Table" };
