// @cobalt/css-popup/src/storybook.js
import React from "react";
import "./styles.scss";

const Content = ({ children }) => (
  <div
    style={{
      display: "flex",
      justifyContent: "center",
      alignItems: "center",
      minHeight: "200px"
    }}
  >
    {children}
  </div>
);

export const usage = () => {
  return (
    <div className="co-popup co-popup--visible">
      <Content>Popup</Content>
    </div>
  );
};

export const sizes = () => {
  return (
    <div style={{ display: "flex" }}>
      <div
        className="co-popup co-popup--visible co-popup--small"
        style={{ marginRight: "20px" }}
      >
        <Content>Small Popup</Content>
      </div>

      <div
        className="co-popup co-popup--visible co-popup--medium"
        style={{ marginRight: "20px" }}
      >
        <Content>Medium Popup</Content>
      </div>

      <div className="co-popup co-popup--visible co-popup--large">
        <Content>Large Popup</Content>
      </div>
    </div>
  );
};

export const compact = () => {
  return (
    <div style={{ width: "0px" }}>
      <div className="co-popup co-popup--visible co-popup--compact">
        Compact Popup
      </div>
    </div>
  );
};

export const variations = () => {
  return (
    <div style={{ display: "flex" }}>
      <div
        className="co-popup co-popup--visible co-popup--small co-popup--light"
        style={{ marginRight: "20px" }}
      >
        <Content>Light Popup</Content>
      </div>
      <div
        className="co-popup co-popup--visible co-popup--small co-popup--dark"
        style={{ marginRight: "20px" }}
      >
        <Content>Dark Popup</Content>
      </div>
      <div className="co-popup co-popup--visible co-popup--small co-popup--error">
        <Content>Error Popup</Content>
      </div>
    </div>
  );
};

export default { title: "Popup" };
