// @cobalt/css-input/src/storybook.js
import React from "react";
import "./styles.scss";

const createMarkup = inputType => {
  const title = inputType.charAt(0).toUpperCase() + inputType.slice(1);

  return (
    <>
      <label
        htmlFor={inputType}
        style={{ display: "block", marginBottom: "4px" }}
      >
        {title}
      </label>
      <input
        className="co-input"
        type={inputType}
        id={inputType}
        style={{ marginBottom: "16px" }}
      />
    </>
  );
};

export const basic = () => {
  const types = ["password", "number", "email", "text"];
  let index = types.length - 1;
  let result = [];

  for (; index >= 0; index--) {
    result.push(createMarkup(types[index]));
  }

  return result;
};

export const sizes = () => {
  return (
    <>
      <label htmlFor="small" style={{ display: "block", marginBottom: "4px" }}>
        Small
      </label>
      <input className="co-input co-input--small" id="small" />
      <br />
      <label htmlFor="medium" style={{ display: "block", marginBottom: "4px" }}>
        Medium
      </label>
      <input className="co-input co-input--medium" id="medium" />
    </>
  );
};

export const variations = () => {
  return (
    <>
      <label htmlFor="error" style={{ display: "block", marginBottom: "4px" }}>
        Error
      </label>
      <input className="co-input co-input--error" id="error" />
      <label
        htmlFor="warning"
        style={{ display: "block", marginTop: "16px", marginBottom: "4px" }}
      >
        Warning
      </label>
      <input className="co-input co-input--warning" id="warning" />
      <label
        htmlFor="unstyled"
        style={{ display: "block", marginBottom: "4px", marginTop: "16px" }}
      >
        Unstyled
      </label>
      <input
        placeholder="Lorem ipsum dolor"
        className="co-input co-input--unstyled"
        id="unstyled"
      />
    </>
  );
};

export const disabled = () => {
  return (
    <>
      <label
        htmlFor="disabled"
        style={{ display: "block", marginBottom: "4px" }}
      >
        Disabled
      </label>
      <input disabled className="co-input" id="disabled" />

      <label
        htmlFor="readonly"
        style={{ display: "block", marginBottom: "4px" }}
      >
        Readonly
      </label>
      <input readOnly className="co-input" value="Some value" id="readonly" />
    </>
  );
};

export const placeholder = () => {
  return (
    <>
      <label
        htmlFor="placeholder"
        style={{ display: "block", marginBottom: "4px" }}
      >
        Placeholder
      </label>
      <input
        placeholder="Lorem ipsum dolor"
        className="co-input"
        id="placeholder"
      />
    </>
  );
};

export const wrap = () => {
  return (
    <div className="co-input co-input--wrap">
      <textarea
        style={{
          height: "50px",
          margin: 4,
          width: "100%",
          resize: "vertical"
        }}
      />
    </div>
  );
};

export const clickable = () => {
  return <input className="co-input co-input--clickable" id="clickable" />;
};

export const alignments = () => {
  return (
    <>
      <label htmlFor="align-left">Align left</label>
      <input
        id="align-left"
        className="co-input co-input--align-left"
        style={{ marginBottom: 16 }}
      />
      <label htmlFor="align-center">Align center</label>
      <input
        id="align-center"
        className="co-input co-input--align-center"
        style={{ marginBottom: 16 }}
      />
      <label htmlFor="align-right">Align right </label>
      <input id="align-right" className="co-input co-input--align-right" />
    </>
  );
};

export default { title: "Input" };
