// @cobalt/css-box/src/storybook.js
import React from "react";
import "./styles.scss";

const Square = ({ children, small }) => (
  <div
    style={{
      width: small ? 50 : 100,
      height: small ? 50 : 100,
      backgroundColor: "#EEE",
      boxSizing: "border-box",
      border: "1px solid #AAA"
    }}
  >
    {children}
  </div>
);

export const primary = () => {
  return (
    <div className="co-box">
      <Square />
      <Square />
    </div>
  );
};

export const padding = () => {
  return (
    <>
      <div className="co-box co-box--padding-2">
        <Square />
      </div>
      <div className="co-box co-box--padding-x-4 co-box--padding-y-4">
        <Square />
      </div>
      <div className="co-box co-box--padding-left-6 co-box--padding-right-6 co-box--padding-top-6 co-box--padding-bottom-6">
        <Square />
      </div>
      <div className="co-box co-box--padding-top-4 co-box--padding-y-6 co-box--padding-2">
        <Square />
      </div>
    </>
  );
};

export const growing = () => {
  return (
    <div style={{ display: "flex" }}>
      <div
        className="co-box co-box--grow"
        style={{ backgroundColor: "#F7F7F7" }}
      >
        <Square />
      </div>
    </div>
  );
};

export default { title: "Box" };
