// @cobalt/css-divider/src/storybook.js
import React from "react";
import "./styles.scss";

export const horizontal = () => {
  return (
    <div>
      <p style={{ margin: "8px 0" }}>
        Cobalt is the chemical element of atomic number 27.
      </p>
      <div className="co-divider"></div>
      <p style={{ margin: "8px 0" }}>
        It is a hard silvery-white magnetic metal.
      </p>
    </div>
  );
};

export const horizontalWithLabel = () => {
  return (
    <div>
      <p style={{ margin: "8px 0" }}>Cobalt 1</p>
      <div className="co-divider">
        <span style={{ margin: "0 8px" }}>OR</span>
      </div>
      <p style={{ margin: "8px 0" }}>Cobalt 2</p>
    </div>
  );
};

export const vertical = () => {
  return (
    <div
      style={{
        height: "300px",
        display: "flex",
        alignItems: "center",
        flexDirection: "row"
      }}
    >
      <p style={{ margin: "0 8px" }}>
        Cobalt is the chemical element of atomic number 27.
      </p>
      <div className="co-divider co-divider--vertical"></div>
      <p style={{ margin: "0 8px" }}>
        It is a hard silvery-white magnetic metal.
      </p>
    </div>
  );
};

export const verticalWithLabel = () => {
  return (
    <div
      style={{
        height: "300px",
        display: "flex",
        alignItems: "center",
        flexDirection: "row"
      }}
    >
      <p style={{ margin: "0 8px" }}>
        Cobalt is the chemical element of atomic number 27.
      </p>
      <div className="co-divider co-divider--vertical">
        <span style={{ margin: "8px 0" }}>OR</span>
      </div>
      <p style={{ margin: "0 8px" }}>
        It is a hard silvery-white magnetic metal.
      </p>
    </div>
  );
};

export const withFlexParent = () => {
  return (
    <>
      <h1> Horizontal: </h1>
      <div
        style={{
          display: "flex",
          flexDirection: "column",
          height: "300px"
        }}
      >
        <p style={{ margin: "0 8px" }}>
          Cobalt is the chemical element of atomic number 27.
        </p>
        <div className="co-divider">
          <span style={{ margin: "8px 0" }}>OR</span>
        </div>
        <p style={{ margin: "0 8px" }}>
          It is a hard silvery-white magnetic metal.
        </p>
      </div>

      <h1> Vertical: </h1>
      <div
        style={{
          height: "300px",
          display: "flex",
          alignItems: "center",
          flexDirection: "row"
        }}
      >
        <p style={{ margin: "0 8px" }}>
          Cobalt is the chemical element of atomic number 27.
        </p>
        <div className="co-divider co-divider--vertical" />
        <p style={{ margin: "0 8px" }}>
          It is a hard silvery-white magnetic metal.
        </p>
      </div>
    </>
  );
};

withFlexParent.story = {
  parameters: {
    docs: { disable: true }
  }
};

export const withBlockParent = () => {
  return (
    <>
      <h1>Horizontal:</h1>
      <div
        style={{
          display: "block",
          height: "300px"
        }}
      >
        <p style={{ margin: "0 8px" }}>
          Cobalt is the chemical element of atomic number 27.
        </p>
        <div className="co-divider">
          <span style={{ margin: "8px 0" }}>OR</span>
        </div>
        <p style={{ margin: "0 8px" }}>
          It is a hard silvery-white magnetic metal.
        </p>
      </div>

      <h1>Vertical:</h1>
      <div
        style={{
          display: "block",
          height: "300px",
          marginTop: "40px"
        }}
      >
        <p style={{ display: "inline-block", margin: "0 8px" }}>
          Cobalt is the chemical element of atomic number 27.
        </p>

        <div className="co-divider co-divider--vertical" />
        <p style={{ display: "inline-block", margin: "0 8px" }}>
          It is a hard silvery-white magnetic metal.
        </p>
      </div>
    </>
  );
};

withBlockParent.story = {
  parameters: {
    docs: { disable: true }
  }
};

export default { title: "Divider" };
