// @cobalt/css-grid/src/storybook.js
import React from "react";
import "./styles.scss";

export const primaryStory = () => {
  return (
    <>
      <h2>The following is a Grid with 2 columns:</h2>
      <div className="co-grid co-grid--2-columns">
        <div
          className="co-grid__column"
          style={{ backgroundColor: "#b8d0f2", border: "1px solid orangered" }}
        >
          Column 1
        </div>
        <div
          className="co-grid__column"
          style={{ backgroundColor: "#b8d0f2", border: "1px solid orangered" }}
        >
          Column 2
        </div>
      </div>
    </>
  );
};

export const numberColumns = () => {
  return (
    <>
      <h2>The following is a Grid with 3 columns per row:</h2>
      <div className="co-grid co-grid--3-columns">
        <div
          className="co-grid__column"
          style={{ backgroundColor: "#b8d0f2", border: "1px solid orangered" }}
        >
          Column 1
        </div>
        <div
          className="co-grid__column"
          style={{ backgroundColor: "#b8d0f2", border: "1px solid orangered" }}
        >
          Column 2
        </div>
        <div
          className="co-grid__column"
          style={{ backgroundColor: "#b8d0f2", border: "1px solid orangered" }}
        >
          Column 3
        </div>
        <div
          className="co-grid__column"
          style={{ backgroundColor: "#b8d0f2", border: "1px solid orangered" }}
        >
          Column 4
        </div>
      </div>
    </>
  );
};

export const padding = () => (
  <div
    className="co-grid co-grid--padding-2 co-grid--padding-y-4 co-grid--padding-top-8"
    style={{ backgroundColor: "#EEE" }}
  >
    <div
      className="co-grid__column"
      style={{ backgroundColor: "#b8d0f2", border: "1px solid orangered" }}
    >
      Column 1
    </div>
  </div>
);

export const gap = () => {
  return (
    <>
      <h2>The following is a Grid with a bigger gap between columns:</h2>
      <div className="co-grid co-grid--2-columns co-grid--gap-8">
        <div
          className="co-grid__column"
          style={{ backgroundColor: "#b8d0f2", border: "1px solid orangered" }}
        >
          Column 1
        </div>
        <div
          className="co-grid__column"
          style={{ backgroundColor: "#b8d0f2", border: "1px solid orangered" }}
        >
          Column 2
        </div>
        <div
          className="co-grid__column"
          style={{ backgroundColor: "#b8d0f2", border: "1px solid orangered" }}
        >
          Column 3
        </div>
      </div>
    </>
  );
};

export const columnLength = () => {
  return (
    <>
      <h2>
        The following is a Grid with space for three columns. The first takes
        one space and the second takes two spaces:
      </h2>
      <div className="co-grid co-grid--3-columns">
        <div
          className="co-grid__column"
          style={{ backgroundColor: "#b8d0f2", border: "1px solid orangered" }}
        >
          Column 1 — default length
        </div>
        <div
          className="co-grid__column co-grid__column--length-2"
          style={{ backgroundColor: "#b8d0f2", border: "1px solid orangered" }}
        >
          Column 2 — lenght = 2
        </div>
      </div>
    </>
  );
};

export const columnPosition = () => {
  return (
    <>
      <h2>
        The following is a Grid with space for three columns. The second column
        is positioned in the space reserved for the third:
      </h2>
      <div className="co-grid co-grid--3-columns">
        <div
          className="co-grid__column"
          style={{ backgroundColor: "#b8d0f2", border: "1px solid orangered" }}
        >
          Column 1 — default position
        </div>
        <div
          className="co-grid__column co-grid__column--position-3"
          style={{ backgroundColor: "#b8d0f2", border: "1px solid orangered" }}
        >
          Column 2 — position = 3
        </div>
      </div>
    </>
  );
};

export default { title: "Grid" };
