// @cobalt/react-box/src/Box.tsx
import React, { FC, CSSProperties, Ref, HTMLAttributes } from "react";
import cc from "classcat";
import styles from "@cobalt/css-box";
import useResponsiveValue, {
  ResponsiveValue,
} from "@cobalt/react-responsive-value-hook";

type Styles = typeof styles;

type As = keyof Pick<
  JSX.IntrinsicElements,
  | "section"
  | "header"
  | "main"
  | "footer"
  | "div"
  | "aside"
  | "article"
  | "details"
  | "nav"
  | "span"
>;

type Space =
  | 0
  | "0"
  | 1
  | "1"
  | 2
  | "2"
  | 3
  | "3"
  | 4
  | "4"
  | 5
  | "5"
  | 6
  | "6"
  | 7
  | "7"
  | 8
  | "8"
  | 9
  | "9"
  | 10
  | "10"
  | 11
  | "11"
  | 12
  | "12";

export type BoxType = any;

export interface Props extends HTMLAttributes<HTMLElement> {
  /** Renders the component using a different element. */
  as?: As;
  backgroundColor?: CSSProperties["backgroundColor"];
  /** Fills the available space in a flex container. */
  grow?: boolean;
  /**
   * Defines the height of the element.
   * Use any valid value for CSS property, or a tuple with values for the supported viewports.
   */
  height?: ResponsiveValue<CSSProperties["height"]>;
  /**
   * Defines the padding.
   * Use values from 0 to 12, or a tuple with values for the supported viewports.
   * These values will be multiplied by a predefined space unit.
   */
  padding?: ResponsiveValue<Space>;
  /**
   * defines the horizontal padding.
   * use values from 0 to 12, or a tuple with values for the supported viewports.
   * these values will be multiplied by a predefined space unit.
   */
  paddingX?: ResponsiveValue<Space>;
  /**
   * defines the vertical padding.
   * use values from 0 to 12, or a tuple with values for the supported viewports.
   * these values will be multiplied by a predefined space unit.
   */
  paddingY?: ResponsiveValue<Space>;
  /**
   * Defines the bottom padding.
   * Use values from 0 to 12, or a tuple with values for the supported viewports.
   * These values will be multiplied by a predefined space unit.
   */
  paddingBottom?: ResponsiveValue<Space>;
  /**
   * Defines the left padding.
   * Use values from 0 to 12, or a tuple with values for the supported viewports.
   * These values will be multiplied by a predefined space unit.
   */
  paddingLeft?: ResponsiveValue<Space>;
  /**
   * Defines the right padding.
   * Use values from 0 to 12, or a tuple with values for the supported viewports.
   * These values will be multiplied by a predefined space unit.
   */
  paddingRight?: ResponsiveValue<Space>;
  /**
   * Defines the top padding.
   * Use values from 0 to 12, or a tuple with values for the supported viewports.
   * These values will be multiplied by a predefined space unit.
   */
  paddingTop?: ResponsiveValue<Space>;
  /** Makes the container scrollable. */
  scrollable?: boolean;
  /**
   * Defines the width of the element.
   * Use any valid value for CSS property, or a tuple with values for the supported viewports.
   */
  width?: ResponsiveValue<CSSProperties["width"]>;
  forwardedRef?: Ref<BoxType>;
}

const Box: FC<Props> = ({
  className = "",
  forwardedRef,
  style,
  as = "div",
  backgroundColor,
  width,
  grow,
  height,
  padding,
  paddingBottom,
  paddingLeft,
  paddingRight,
  paddingTop,
  paddingX,
  paddingY,
  scrollable,
  ...props
}) => {
  const Tag = as;
  const rWidth = useResponsiveValue(width);
  const rHeight = useResponsiveValue(height);
  const rPadding = useResponsiveValue(padding);
  const rPaddingX = useResponsiveValue(paddingX);
  const rPaddingY = useResponsiveValue(paddingY);
  const rPaddingBottom = useResponsiveValue(paddingBottom);
  const rPaddingLeft = useResponsiveValue(paddingLeft);
  const rPaddingRight = useResponsiveValue(paddingRight);
  const rPaddingTop = useResponsiveValue(paddingTop);

  const classNames = cc([
    styles["co-box"],
    {
      [styles["co-box--scrollable"]]: scrollable,
      [styles["co-box--grow"]]: grow,
      [styles[`co-box--padding-${rPadding}` as keyof Styles]]:
        rPadding !== undefined,
      [styles[`co-box--padding-x-${rPaddingX}` as keyof Styles]]:
        rPaddingX !== undefined,
      [styles[`co-box--padding-y-${rPaddingY}` as keyof Styles]]:
        rPaddingY !== undefined,
      [styles[`co-box--padding-bottom-${rPaddingBottom}` as keyof Styles]]:
        rPaddingBottom !== undefined,
      [styles[`co-box--padding-left-${rPaddingLeft}` as keyof Styles]]:
        rPaddingLeft !== undefined,
      [styles[`co-box--padding-right-${rPaddingRight}` as keyof Styles]]:
        rPaddingRight !== undefined,
      [styles[`co-box--padding-top-${rPaddingTop}` as keyof Styles]]:
        rPaddingTop !== undefined,
    },
    className,
  ]);

  return (
    <Tag
      className={classNames}
      ref={forwardedRef}
      style={{
        width: rWidth,
        height: rHeight,
        backgroundColor,
        ...style,
      }}
      {...props}
    />
  );
};

export default Box;
