// @cobalt/css-textarea/src/storybook.js
import React from "react";
import "./styles.scss";

export const basic = () => {
  return (
    <>
      <textarea className="co-textarea" placeholder="Write something" />
      <textarea
        className="co-textarea"
        rows={5}
        placeholder="Write something"
      />
    </>
  );
};

export const sizes = () => {
  return (
    <>
      <textarea
        className="co-textarea co-textarea--small"
        placeholder="Write something"
      />
      <textarea
        className="co-textarea co-textarea--medium"
        placeholder="Write something"
      />
    </>
  );
};

export const error = () => {
  return (
    <textarea
      className="co-textarea co-textarea--error"
      placeholder="Write something"
    />
  );
};

export const warning = () => {
  return (
    <textarea
      className="co-textarea co-textarea--warning"
      placeholder="Write something"
    />
  );
};

export const unstyled = () => {
  return (
    <textarea
      className="co-textarea co-textarea--unstyled"
      placeholder="Write something"
    />
  );
};

export const disabled = () => {
  return (
    <textarea className="co-textarea" placeholder="Write something" disabled />
  );
};

export const readonly = () => {
  return (
    <textarea
      className="co-textarea"
      defaultValue="This is a default value "
      readOnly
    />
  );
};

export const nonResizable = () => {
  return (
    <textarea
      className="co-textarea co-textarea--non-resizable"
      placeholder="Write something"
    />
  );
};

export default { title: "Textarea" };
